site/using-smart-accounts/estimate-gas/estimate-user-op-gas-state-overrides.md

```diff
@@ -0,0 +1,121 @@
+
+### Setting up as a ward through state overrides
+
+To estimate gas for the user operation transaction, you need to simulate a transaction where you mint DAI. This requires overriding the state to make our address appear as a ward. In other words, you need to pass in the state overrides value, along with the user operation transaction data, to adjust the wards map in the DAI contract to include your address with the necessary permissions.
+
+The state override is a mapping of specific storage locations in the smart contract as keys of the map, pointing to the override values for the simulation.
+
+```ts
+{
+  stateOverrides: {
+    "0xDAI_CONTRACT_ADDRESS": {
+      storage: {
+        "0xSTORAGE_SLOT_OF_YOUR_ADDRESS": "0x0000000000000000000000000000000000000000000000000000000000000001"
+      }
+    }
+  }
+}
+```
+
+::: tip Storage Slot Calculation
+Calculating the storage location depends on the structure of the contract’s storage and the specific variable you want to override. For the DAI contract example, the storage location for the wards mapping for your address is calculated based on the [Solidity storage layout rules](https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#mappings-and-dynamic-arrays):
+
+```ts
+const yourAddress = "0x...";
+const overrideStorageLocation = keccak256(
+  concatenate(hexZeroPad(yourAddress), hexZeroPad(0x0))
+); // 0xSTORAGE_SLOT_OF_YOUR_ADDRESS
+```
+
+:::
+
+## Simulated user operation gas estimation using `SmartAccountClient`
+
+[`eth_estimateUserOperationGas`](https://docs.alchemy.com/reference/eth-estimateuseroperationgas) is an RPC method that bundlers support as per the [`ERC-4337`](https://eips.ethereum.org/EIPS/eip-4337#-eth_estimateuseroperationgas) specification. It estimate the gas values for a UserOperation. Given a UserOperation, optionally without gas limit or price fields, this method returns the needed gas limits.
+
+`SmartAccountClient` of `aa-sdk` provides default [`gasEstimator`](/packages/aa-core/smart-account-client/middleware/index#gasEstimator) that internally includes calling `eth_estimateUserOperationGas` to the bundler in addition to applying user operation overrides or fee options for populating the user operation gas fields in a most desired mannder. If you are looking to estimate gas of a user operation without building the entire user operation through the middleware pipeline, you can call the [`estimateUserOperationGas`](/packages/aa-core/smart-account-client/actions/estimateUserOperationGas) action on the `SmartAccountClient` to directly fetch network user operation gas estimates from the bundler. The action returns `UserOperationEstimateGasResponse` containing the estimated gas values.
+
+::: tip Note 1
+The actual gas estimation or fee estimation is performed by the bundler connected to the `SmartAccountClient`, so depending on the bundler you are using, the gas estimate value might be different.
```

Comment:
(Minor) I find the references to fee estimation in this paragraph and the one above a little bit confusing, since `estimateUserOperationGas` does not provide fee estimation. I might take them out, or if you want to keep them around then clarify somewhere that `estimateUserOperationGas` estimates gas quantities but not fees.